### ISSUE TITLE
Scope leak of comprehension variable after an `isintance` check

### ISSUE THREAD

**Issue by jfmend** [microsoft/pyright#10301]
## Environment data
Tested on:
- pylance v2025.4.1 
- python 3.10.12

## Code Snippet

```python
def main() -> None:
    print([letter
        for letter in ["a", 1]
        if  isinstance(letter, str)])
    letter() # <- False positive: letter is inferred to be of type: (function) letter: (() -> None) | str

def letter() -> None:
    pass
```

The type hint for `letter` at the last statement of main is: `(function) letter: (() -> None) | str`

## Expected behavior

No errors.
The type hint for `letter` at the last statement of main should be: `(function) letter: (() -> None)`

## Actual behavior

Since the type of `letter` is inferred to be `(() -> None) | str`, pyright gives the following error message:

```
Object of type "str" is not callable
  Attribute "__call__" is unknown
```

This error only happens if the comprehension has an if clause with an `isinstance` check.
After the statement containing the comprehension, the scope of the loop control variable leaks out with whatever type `isinstance` checked for.

**Comment by erictraut** [comment]
This is addressed in 1.1.400.
